### ISSUE TITLE
Unable to collect docstrings from `.pyi` files

### ISSUE THREAD

**Issue by samuelcolvin** [mkdocstrings/mkdocstrings#404]
(not sure which repo this goes in, so creating it here for simplicity.)

Sorry, it's me again trying to push the limits of your amazing library.

## Describe the bug

I have a library that's partly implemented in rust and I want to build docs from docstrings defined in a `.pyi` files which defines the python interface for the rust code.

## To Reproduce

See [`_rust_notify.pyi`](https://github.com/samuelcolvin/watchfiles/blob/main/watchfiles/_rust_notify.pyi), and the PR where tried to add docs from that file https://github.com/samuelcolvin/watchfiles/pull/104.

Currently I'm getting the following error while trying to build with a reference to the class in that file:

```
INFO     -  [14:19:37] Detected file changes
INFO     -  Building documentation...
INFO     -  running examples tests...
..............                                                                                                                                                                                                      [100%]
14 passed in 0.07s
WARNING  -  mkdocstrings_handlers: 1 aliases were still unresolved after 2 iterations
ERROR    -  mkdocstrings: Template 'alias.html' not found for 'python' handler and theme 'material'.
ERROR    -  Error reading page 'api/rust_backend.md': alias.html
alias.html
```

The `mkdocstrings_handlers: 1 aliases were still unresolved after 2 iterations` line refers to the following errors in verbose output:

```
DEBUG    -  griffe: Found watchfiles: loading
DEBUG    -  griffe: Loading path watchfiles/__init__.py
DEBUG    -  griffe: Loading path watchfiles/version.py
DEBUG    -  griffe: Loading path watchfiles/filters.py
DEBUG    -  griffe: Loading path watchfiles/cli.py
DEBUG    -  griffe: Loading path watchfiles/main.py
DEBUG    -  griffe: Loading path watchfiles/__main__.py
DEBUG    -  griffe: Loading path watchfiles/_rust_notify.cpython-310-x86_64-linux-gnu.so
DEBUG    -  griffe: Alias watchfiles.BaseFilter was resolved to watchfiles.filters.BaseFilter
DEBUG    -  griffe: Alias watchfiles.DefaultFilter was resolved to watchfiles.filters.DefaultFilter
DEBUG    -  griffe: Alias watchfiles.PythonFilter was resolved to watchfiles.filters.PythonFilter
DEBUG    -  griffe: Alias watchfiles.Change was resolved to watchfiles.main.Change
DEBUG    -  griffe: Alias watchfiles.arun_process was resolved to watchfiles.main.arun_process
DEBUG    -  griffe: Alias watchfiles.awatch was resolved to watchfiles.main.awatch
DEBUG    -  griffe: Alias watchfiles.run_process was resolved to watchfiles.main.run_process
DEBUG    -  griffe: Alias watchfiles.watch was resolved to watchfiles.main.watch
DEBUG    -  griffe: Alias watchfiles.VERSION was resolved to watchfiles.version.VERSION
DEBUG    -  griffe: Alias resolution error for watchfiles._rust_notify.RustNotify -> builtins.RustNotify
DEBUG    -  griffe: Iteration 1 finished, 9 aliases resolved, still 1 to go
DEBUG    -  griffe: Alias resolution error for watchfiles._rust_notify.RustNotify -> builtins.RustNotify
DEBUG    -  griffe: Iteration 2 finished, 0 aliases resolved, still 1 to go
WARNING  -  mkdocstrings_handlers: 1 aliases were still unresolved after 2 iterations
```

Basically the same problem, but as a warning

## Expected behavior

Since these files have python syntax, I imagine it might be quite easy to extract docstrings?

## Information (please complete the following information):
- OS: [e.g. iOS] Ubuntu
- Browser: [e.g. chrome, safari] NA
- `mkdocstrings` version: [e.g. 0.10.2] 

```
black==21.12b0
mkdocs==1.2.3
mkdocs-material==8.2.6
mkdocs-simple-hooks==0.1.5
mkdocstrings[python]==0.18.1
```

**Comment by samuelcolvin** [comment]
Sorry,  I had the wrong error before due to a mistake my side, updated.

**Comment by pawamoy** [comment]
> Sorry, it's me again trying to push the limits of your amazing library.

Fine by me :joy: 

Indeed, support for `.pyi` files would be great. I believe it shouldn't be that hard to implement since, as you stated, `.pyi` files are using regular Python syntax. Needs a bit of thought of course.

**Comment by pawamoy** [comment]
Version 0.19.0 of Griffe is able to both load `*.pyi` stubs and merge the data back into their existing, counter-part `*.py` modules. Let me know if it works for you! Can reopen if it does not.

**Comment by samuelcolvin** [comment]
Thanks so much. I'll check, there's no .py file though, just a .so shared library.

**Comment by samuelcolvin** [comment]
Thanks so much, this is now (sort of) working, see https://github.com/samuelcolvin/watchfiles/pull/165. The next release of watchfiles will use this to generate docs for the rust interface.

However I've discovered a small bug with this: 

watchfiles is now built using abi3 which creates a `_rust_notify.abi3.so` file alongside `_rust_notify.pyi`. If the `_rust_notify.abi3.so` file is present, and (I guess) dependencies are installed so mkdocstrings can try to load the package, I get the same error as before:

```
jinja2.exceptions.TemplateNotFound: alias.html
```

If `_rust_notify.abi3.so` is deleted or dependencies are not installed, `mkdocs build` works fine.

**Comment by pawamoy** [comment]
Hey @samuelcolvin, you might have seen from #451 that this error is triggered because an alias in your code cannot be resolved, and this in turn is triggered by the fact a compiled object tells its parent module is `builtins` (see https://github.com/oconnor663/blake3-py/issues/38#issuecomment-1353810913).  A probable solution is that you fix that the same way you suggested in the mentioned issue. I'll try again running mkdocstrings/Griffe on your code anyway, there were other bugfixes since then.

**Comment by samuelcolvin** [comment]
That's really helpful, thanks so much.

**Comment by pawamoy** [comment]
Take care to use the canonical path of the parent module when setting the `__module__` attribute, i.e. the module where the object is actually defined within the whole tree of the package (not just the compiled modules), and not where the object is imported, even if this is the public path to the object.

**Comment by samuelcolvin** [comment]
Yup, it's fixed in pydantic-core, I'll fix it in watchfiles soon.

**Comment by ZhiyuanChen** [comment]
Can I ask how to get enable pyi? 

I have the following [doc page](https://github.com/ZhiyuanChen/open-feishu/blob/master/docs/docs/api/im/messages.md):

```markdown
::: feishu.api.im.messages
```

And both [`.py`](https://github.com/ZhiyuanChen/open-feishu/blob/master/feishu/api/im/messages.py) and [`.pyi`](https://github.com/ZhiyuanChen/open-feishu/blob/master/feishu/api/im/messages.pyi) file. 

But doc page only shows docstrings from `.py` file

http://feishu.danling.org/api/im/messages/

**Comment by pawamoy** [comment]
@ZhiyuanChen can you open a new issue please? In there, please also share what you see and what you would expect to see instead.

**Comment by ZhiyuanChen** [comment]
> @ZhiyuanChen can you open a new issue please? In there, please also share what you see and what you would expect to see instead.

Thank you so much! I'll do that once I get back to work on Wednesday
